File: apple/Sources/Autograph/KeyPair.swift
import Clibautograph
import Foundation

func createIdentityKeyPair() -> [UInt8] {
    createBytes(autograph_identity_key_pair_size())
}

func createSessionKeyPair() -> [UInt8] {
    createBytes(autograph_session_key_pair_size())
}

func createIdentityPublicKey() -> [UInt8] {
    createBytes(autograph_identity_public_key_size())
}

func createSessionPublicKey() -> [UInt8] {
    createBytes(autograph_session_public_key_size())
}

public func generateIdentityKeyPair() throws -> [UInt8] {
    var keyPair = createIdentityKeyPair()
    let success = autograph_identity_key_pair(&keyPair)
    if !success {
        throw AutographError.keyPair
    }
    return keyPair
}

public func generateSessionKeyPair() throws -> [UInt8] {
    var keyPair = createSessionKeyPair()
    let success = autograph_session_key_pair(&keyPair)
    if !success {
        throw AutographError.keyPair
    }
    return keyPair
}

public func getIdentityPublicKey(_ keyPair: [UInt8]) -> [UInt8] {
    var publicKey = createIdentityPublicKey()
    autograph_get_identity_public_key(&publicKey, keyPair)
    return publicKey
}

public func getSessionPublicKey(_ keyPair: [UInt8]) -> [UInt8] {
    var publicKey = createSessionPublicKey()
    autograph_get_session_public_key(&publicKey, keyPair)
    return publicKey
}

public func getPublicKeys(
    _ identityKeyPair: [UInt8],
    _ sessionKeyPair: [UInt8]
) -> ([UInt8], [UInt8]) {
    let identityKey = getIdentityPublicKey(identityKeyPair)
    let sessionKey = getSessionPublicKey(sessionKeyPair)
    return (identityKey, sessionKey)
}

File: android/src/main/java/sh/autograph/KeyPair.kt
package sh.autograph

internal class KeyPair {
    companion object {
        init {
            System.loadLibrary("autograph")
        }

        private external fun autographIdentityKeyPairSize(): Int

        private external fun autographSessionKeyPairSize(): Int

        private external fun autographIdentityPublicKeySize(): Int

        private external fun autographSessionPublicKeySize(): Int

        private external fun autographIdentityKeyPair(keyPair: ByteArray): Boolean

        private external fun autographSessionKeyPair(keyPair: ByteArray): Boolean

        private external fun autographGetIdentityPublicKey(
            publicKey: ByteArray,
            keyPair: ByteArray,
        )

        private external fun autographGetSessionPublicKey(
            publicKey: ByteArray,
            keyPair: ByteArray,
        )

        fun createIdentityKeyPair(): ByteArray = ByteArray(autographIdentityKeyPairSize())

        fun createSessionKeyPair(): ByteArray = ByteArray(autographSessionKeyPairSize())

        fun createIdentityPublicKey(): ByteArray = ByteArray(autographIdentityPublicKeySize())

        fun createSessionPublicKey(): ByteArray = ByteArray(autographSessionPublicKeySize())

        fun generateIdentityKeyPair(): ByteArray {
            val keyPair = createIdentityKeyPair()
            val success = autographIdentityKeyPair(keyPair)
            if (!success) {
                throw RuntimeException("Key generation failed")
            }
            return keyPair
        }

        fun generateSessionKeyPair(): ByteArray {
            val keyPair = createSessionKeyPair()
            val success = autographSessionKeyPair(keyPair)
            if (!success) {
                throw RuntimeException("Key generation failed")
            }
            return keyPair
        }

        fun getIdentityPublicKey(keyPair: ByteArray): ByteArray {
            val publicKey = createIdentityPublicKey()
            autographGetIdentityPublicKey(publicKey, keyPair)
            return publicKey
        }

        fun getSessionPublicKey(keyPair: ByteArray): ByteArray {
            val publicKey = createSessionPublicKey()
            autographGetSessionPublicKey(publicKey, keyPair)
            return publicKey
        }

        fun getPublicKeys(
            identityKeyPair: ByteArray,
            sessionKeyPair: ByteArray,
        ): Pair<ByteArray, ByteArray> {
            val identityKey = getIdentityPublicKey(identityKeyPair)
            val sessionKey = getSessionPublicKey(sessionKeyPair)
            return Pair(identityKey, sessionKey)
        }
    }
}

public fun generateIdentityKeyPair(): ByteArray {
    return KeyPair.generateIdentityKeyPair()
}

public fun generateSessionKeyPair(): ByteArray {
    return KeyPair.generateSessionKeyPair()
}

public fun getIdentityPublicKey(keyPair: ByteArray): ByteArray {
    return KeyPair.getIdentityPublicKey(keyPair)
}

public fun getSessionPublicKey(keyPair: ByteArray): ByteArray {
    return KeyPair.getSessionPublicKey(keyPair)
}

public fun getPublicKeys(
    identityKeyPair: ByteArray,
    sessionKeyPair: ByteArray,
): Pair<ByteArray, ByteArray> {
    return KeyPair.getPublicKeys(identityKeyPair, sessionKeyPair)
}
